File: Day27/scion/pkg/ent/schema/agent.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package schema

import (
	"time"

	"entgo.io/ent"
	"entgo.io/ent/schema/edge"
	"entgo.io/ent/schema/field"
	"entgo.io/ent/schema/index"
	"github.com/google/uuid"
)

// Agent holds the schema definition for the Agent entity.
// Only principal-relevant fields are included; operational fields
// (ContainerStatus, RuntimeState, etc.) will be added when the
// agent entity is fully migrated to Ent.
type Agent struct {
	ent.Schema
}

// Fields of the Agent.
func (Agent) Fields() []ent.Field {
	return []ent.Field{
		field.UUID("id", uuid.UUID{}).
			Default(uuid.New).
			Immutable(),
		field.String("slug").
			NotEmpty(),
		field.String("name").
			NotEmpty(),
		field.String("template").
			Optional(),
		field.UUID("grove_id", uuid.UUID{}),
		field.Enum("status").
			Values("created", "provisioning", "cloning", "starting", "running", "stopping", "stopped", "error").
			Default("created"),
		field.UUID("created_by", uuid.UUID{}).
			Optional().
			Nillable(),
		field.UUID("owner_id", uuid.UUID{}).
			Optional().
			Nillable(),
		field.Bool("delegation_enabled").
			Default(false),
		field.String("visibility").
			Default("private"),
		field.Time("created").
			Default(time.Now).
			Immutable(),
		field.Time("updated").
			Default(time.Now).
			UpdateDefault(time.Now),
	}
}

// Edges of the Agent.
func (Agent) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("grove", Grove.Type).
			Ref("agents").
			Field("grove_id").
			Required().
			Unique(),
		edge.From("creator", User.Type).
			Ref("created_agents").
			Field("created_by").
			Unique(),
		edge.From("owner", User.Type).
			Ref("owned_agents").
			Field("owner_id").
			Unique(),
		edge.From("memberships", GroupMembership.Type).
			Ref("agent"),
		edge.From("policy_bindings", PolicyBinding.Type).
			Ref("agent"),
	}
}

// Indexes of the Agent.
func (Agent) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("slug", "grove_id").
			Unique(),
	}
}

File: Day27/scion/examples/adk_scion_agent/agent.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Root agent definition for the ADK scion example.

This module defines an ADK LlmAgent that integrates with scion's lifecycle
management. It bridges auth environment variables, wires up status-reporting
callbacks, and exposes file_write and sciontool_status as callable tools.

When run with `adk run`, the agent operates as an interactive coding assistant
that reports its status to scion throughout its lifecycle.
"""

import os

from google.adk import Agent

from . import callbacks, tools

# ---------------------------------------------------------------------------
# Auth bridging
# ---------------------------------------------------------------------------
# ADK requires GOOGLE_API_KEY for Gemini API access. Scion's Gemini harness
# provides GEMINI_API_KEY instead. Bridge the gap at import time.
if not os.environ.get("GOOGLE_API_KEY") and os.environ.get("GEMINI_API_KEY"):
    os.environ["GOOGLE_API_KEY"] = os.environ["GEMINI_API_KEY"]

# ---------------------------------------------------------------------------
# Model configuration
# ---------------------------------------------------------------------------
MODEL = os.environ.get("ADK_MODEL", "gemini-3-flash-preview")

# ---------------------------------------------------------------------------
# Agent instruction
# ---------------------------------------------------------------------------
AGENT_INSTRUCTION = """\
You are a coding assistant running inside a scion-managed container.

Your workspace is mounted at /workspace (or the current working directory if
running outside a container). You can read, create, and modify files there.

## Available tools

- **file_write(file_path, content)**: Create or overwrite a file in the
  workspace. Paths are relative to the workspace root unless absolute.

- **sciontool_status(status_type, message)**: Signal lifecycle events to scion.
  - Call `sciontool_status("ask_user", "<your question>")` **before** you ask
    the user a question. This lets scion know you are waiting for input.
  - Call `sciontool_status("task_completed", "<summary>")` when you have
    finished the user's task. Summarize what you did.

## Workflow

1. When you receive a task, work through it step by step.
2. Use file_write to create or modify files as needed.
3. If you need clarification, call sciontool_status("ask_user", ...) first,
   then ask your question.
4. When the task is complete, call sciontool_status("task_completed", ...)
   with a brief summary of what you accomplished.
"""

# ---------------------------------------------------------------------------
# Agent construction
# ---------------------------------------------------------------------------
root_agent = Agent(
    model=MODEL,
    name="scion_agent",
    instruction=AGENT_INSTRUCTION,
    tools=[tools.file_write, tools.sciontool_status],
    before_agent_callback=callbacks.before_agent_callback,
    before_tool_callback=callbacks.before_tool_callback,
    after_tool_callback=callbacks.after_tool_callback,
    after_agent_callback=callbacks.after_agent_callback,
)
